### PR TITLE
Particle systems to WebGPU

### PR DESCRIPTION
**Pull request by deltakosh**

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15422/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15422/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15422/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15422/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15422/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15422/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15422/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15422/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15422/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15422/merge/testResults/webgpuplaywright/index.html

File: packages/dev/core/src/ShadersWGSL/particles.fragment.fx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+	var textureColor: vec4f = textureSample(diffuseSampler, diffuseSamplerSampler, input.vUV);
+	var baseColor: vec4f = (textureColor * uniforms.textureMask + ( vec4f(1., 1., 1., 1.) - uniforms.textureMask)) * input.vColor;
+
+	#ifdef RAMPGRADIENT
+		var alpha: f32 = baseColor.a;
+		var remappedColorIndex: f32 = clamp((alpha - input.remapRanges.x) / input.remapRanges.y, 0.0, 1.0);
+
+		var rampColor: vec4f = textureSample(rampSampler, rampSamplerSampler, vec2f(1.0 - remappedColorIndex, 0.));
+		baseColor.rgb *= rampColor.rgb;
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/particles.fragment.fx:43]:
```suggestion
		baseColor = vec4f(baseColor.rgb * rampColor.rgb, baseColor.a);
```

File: packages/dev/core/src/ShadersWGSL/particles.fragment.fx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+		// Remapped alpha
+		var finalAlpha: f32 = baseColor.a;
+		baseColor.a = clamp((alpha * rampColor.a - input.remapRanges.z) / input.remapRanges.w, 0.0, 1.0);
+	#endif
+
+	#ifdef BLENDMULTIPLYMODE
+		var sourceAlpha: f32 = input.vColor.a * textureColor.a;
+		baseColor.rgb = baseColor.rgb * sourceAlpha +  vec3f(1.0) * (1.0 - sourceAlpha);
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/particles.fragment.fx:52]:
```suggestion
		baseColor = vec4f(baseColor.rgb * sourceAlpha +  vec3f(1.0) * (1.0 - sourceAlpha), baseColor.a);
```

File: packages/dev/core/src/ShadersWGSL/particles.vertex.fx [hunk 3]
```diff
@@ -78,93 +78,89 @@ fn rotateAlign(toCamera: vec3f, rotatedCorner: vec3f) -> vec3f {
...
 	// Rotate
 	var rotatedCorner: vec3f;
 
 #ifdef BILLBOARDY
-	rotatedCorner.x = cornerPos.x * cos(vertexInputs.angle) - cornerPos.y * sin(vertexInputs.ngle);
-	rotatedCorner.z = cornerPos.x * sin(vertexInputs.angle) + cornerPos.y * cos(vertexInputs.angle);
+	rotatedCorner.x = cornerPos.x * cos(input.angle) - cornerPos.y * sin(input.angle);
+	rotatedCorner.z = cornerPos.x * sin(input.angle) + cornerPos.y * cos(input.angle);
 	rotatedCorner.y = 0.;
     rotatedCorner.xz += uniforms.translationPivot;
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/particles.vertex.fx:105]:
```suggestion
    rotatedCorner.x += uniforms.translationPivot;
    rotatedCorner.z += uniforms.translationPivot;
```

File: packages/dev/core/src/ShadersWGSL/particles.vertex.fx [hunk 4]
```diff
@@ -78,93 +78,89 @@ fn rotateAlign(toCamera: vec3f, rotatedCorner: vec3f) -> vec3f {
...
 	var rotatedCorner: vec3f;
-	rotatedCorner.x = cornerPos.x * cos(vertexInputs.angle) - cornerPos.y * sin(vertexInputs.angle);
-	rotatedCorner.z = cornerPos.x * sin(vertexInputs.angle) + cornerPos.y * cos(vertexInputs.angle);
+	rotatedCorner.x = cornerPos.x * cos(input.angle) - cornerPos.y * sin(input.angle);
+	rotatedCorner.z = cornerPos.x * sin(input.angle) + cornerPos.y * cos(input.angle);
 	rotatedCorner.y = 0.;
-    rotatedCorner.x += uniforms.translationPivot.x;
-	rotatedCorner.y += uniforms.translationPivot.y;
-
+    rotatedCorner.xz += uniforms.translationPivot;
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/particles.vertex.fx:146]:
```suggestion
    rotatedCorner.x += uniforms.translationPivot;
    rotatedCorner.z += uniforms.translationPivot;
```

**Review by sebavan — APPROVED** [review]

**Pull request merged**